File: BOJ/4008.py
# 64MB -> 1500만 개의 int가 아님
# 총 int 개수 : 500만 개

import sys
input = sys.stdin.readline
from collections import deque

n = int(input())
a, b, c = map(int, input().split())
x = list(map(int, input().split()))

X = [0]
for i in range(n):
    X.append(X[-1] + x[i])

# dp[i]는 i번째 병사까지 배치하는 최대 전투력(i번째에서 끊김)
# dp[i] = max(-2aXj * Xi + (aXj^2 - bXj + dp[j])) + (aXi^2 + bXi + c)
# 기울기 -2aXj는 계속 감소한다
# Xi = x0 + x1 + x2 + ... + xi (x0 = 0)

def meetPoint(line1, line2):
    return float((line2[1] - line1[1]) / (line1[0] - line2[0]))

# dp = [0]
graph = deque([[0, 0, 0]])
index = 0
for i in range(1, n + 1):
    # index를 조절
    while index != len(graph) - 1 and X[i] > graph[index + 1][2]:
        index += 1
    while index > 0:
        graph.popleft()
        index -= 1
    # dp 최신화
    # dp.append(graph[index][0] * X[i] + graph[index][1] + a * X[i] ** 2 + b * X[i] + c)
    # graph 최신화
    while graph[-1][2] >= meetPoint(graph[-1], [(-2) * a * X[i], a * X[i] ** 2 - b * X[i] + graph[index][0] * X[i] + graph[index][1] + a * X[i] ** 2 + b * X[i] + c]):
        if index == len(graph) - 1:
            index -= 1
        graph.pop()
    graph.append([(-2) * a * X[i], a * X[i] ** 2 - b * X[i] + graph[index][0] * X[i] + graph[index][1] + a * X[i] ** 2 + b * X[i] + c,
                  meetPoint(graph[-1], ((-2) * a * X[i], a * X[i] ** 2 - b * X[i] + graph[index][0] * X[i] + graph[index][1] + a * X[i] ** 2 + b * X[i] + c))])
    if i == n:
        print(graph[index][0] * X[i] + graph[index][1] + a * X[i] ** 2 + b * X[i] + c)

File: BOJ/4008.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
LL INF = LLONG_MAX;

int n;
LL a, b, c, x[1000001];

struct Line {
	mutable LL k, m;
    mutable double p;
	bool operator < (const Line& o) const { return k < o.k; }
	bool operator < (LL x) const { return p < x; }
};

struct LineContainer : multiset<Line, less<>> {
	double div(LL a, LL b) {
		return (double) a / b;
    }
	bool isect(iterator x, iterator y) {
		if (y == end()) return x->p = INF, 0;
		if (x->k == y->k) x->p = x->m > y->m ? INF : -INF;
		else x->p = div(y->m - x->m, x->k - y->k);
		return x->p >= y->p;
	}
	void add(LL k, LL m) {
		auto z = insert({k, m, 0}), y = z++, x = y;
		while (isect(y, z)) z = erase(z);
		if (x != begin() && isect(--x, y)) isect(x, y = erase(y));
		while ((y = x) != begin() && (--x)->p >= y->p) isect(x, erase(y));
	}
	LL query(LL x) {
		assert(!empty());
		while (begin()->p < x) erase(begin());
        return begin()->k * x + begin()->m;
	}
};

// dp[i] = max(-2aXj * Xi + (aXj^2 - bXj + dp[j])) + (aXi^2 + bXi + c)
int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> n;
    cin >> a >> b >> c;
    LL temp, dp;
    LineContainer graph;
    graph.add(0, 0);
    for (int i = 1; i <= n; ++i) {
        cin >> temp;
        x[i] = x[i - 1] + temp;
        dp = graph.query(x[i]) + a * x[i] * x[i] + b * x[i] + c;
        graph.add((-2) * a * x[i], a * x[i] * x[i] - b * x[i] + dp);
    }
    cout << dp;
}